File: mimo-voice/api/mimo_tts.py
"""
MiMo Voice — MiMo V2-TTS Client
Replaces gTTS (robotic) with natural MiMo speech synthesis.
"""

import os
import time
import base64
import requests
from typing import Optional, Literal
from dataclasses import dataclass


@dataclass
class TTSResult:
    """Result from TTS synthesis."""
    audio_data: bytes           # Raw audio bytes (WAV/MP3)
    format: str                 # "wav" or "mp3"
    duration_ms: float          # Audio duration in ms
    latency_ms: float           # API latency in ms
    text: str                   # Source text
    language: str               # Language code
    voice: str                  # Voice used


class MiMoTTS:
    """
    MiMo V2-TTS client.
    
    Replaces gTTS with natural, expressive speech synthesis.
    Supports emotion, intonation, and natural pacing.
    """

    # Voice presets by language
    VOICE_PRESETS = {
        "en": {"default": "nova", "male": "echo", "female": "nova"},
        "ja": {"default": "yuki", "male": "haru", "female": "yuki"},
        "zh": {"default": "xiaoyi", "male": "yunxi", "female": "xiaoyi"},
        "ko": {"default": "minji", "male": "hyunwoo", "female": "minji"},
        "es": {"default": "elena", "male": "carlos", "female": "elena"},
        "fr": {"default": "claire", "male": "antoine", "female": "claire"},
        "de": {"default": "hanna", "male": "maximilian", "female": "hanna"},
    }

    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = "https://api.xiaomimimo.com/v1",
        model: str = "mimo-v2-tts",
        timeout: int = 30,
    ):
        self.api_key = api_key or os.environ.get("MIMO_API_KEY")
        if not self.api_key:
            raise ValueError(
                "MiMo API key required. Set MIMO_API_KEY env var or pass api_key.\n"
                "Register at: https://platform.xiaomimimo.com"
            )
        self.base_url = base_url.rstrip("/")
        self.model = model
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        })

    def synthesize(
        self,
        text: str,
        language: str = "en",
        voice: Optional[str] = None,
        speed: float = 1.0,
        emotion: Literal["neutral", "happy", "sad", "urgent", "calm"] = "neutral",
        output_format: Literal["wav", "mp3"] = "wav",
    ) -> TTSResult:
        """
        Synthesize speech from text.
        
        Args:
            text: Text to speak
            language: Language code (ISO 639-1)
            voice: Voice name (auto-selected per language if None)
            speed: Speech speed (0.5-2.0)
            emotion: Emotional tone
            output_format: Audio format
            
        Returns:
            TTSResult with audio data
        """
        if voice is None:
            voice = self._get_default_voice(language)

        start_time = time.time()

        payload = {
            "model": self.model,
            "input": text,
            "voice": voice,
            "speed": speed,
            "response_format": output_format,
            "language": language,
        }

        # Add emotion hint via instruction if not neutral
        if emotion != "neutral":
            emotion_hints = {
                "happy": "Speak with a warm, happy tone.",
                "sad": "Speak with a gentle, somber tone.",
                "urgent": "Speak urgently and clearly.",
                "calm": "Speak calmly and slowly.",
            }
            payload["instruction"] = emotion_hints.get(emotion, "")

        response = self.session.post(
            f"{self.base_url}/audio/speech",
            json=payload,
            timeout=self.timeout,
        )
        response.raise_for_status()

        audio_data = response.content
        latency = (time.time() - start_time) * 1000

        return TTSResult(
            audio_data=audio_data,
            format=output_format,
            duration_ms=self._estimate_duration(audio_data, output_format),
            latency_ms=latency,
            text=text,
            language=language,
            voice=voice,
        )

    def synthesize_streaming(
        self,
        text: str,
        language: str = "en",
        voice: Optional[str] = None,
        chunk_size: int = 4096,
    ):
        """
        Stream audio chunks as they're generated.
        
        Yields audio chunks for real-time playback.
        """
        if voice is None:
            voice = self._get_default_voice(language)

        payload = {
            "model": self.model,
            "input": text,
            "voice": voice,
            "language": language,
            "stream": True,
        }

        response = self.session.post(
            f"{self.base_url}/audio/speech",
            json=payload,
            timeout=self.timeout,
            stream=True,
        )
        response.raise_for_status()

        for chunk in response.iter_content(chunk_size=chunk_size):
            if chunk:
                yield chunk

    def _get_default_voice(self, language: str) -> str:
        """Get default voice for language."""
        presets = self.VOICE_PRESETS.get(language, self.VOICE_PRESETS["en"])
        return presets["default"]

    @staticmethod
    def _estimate_duration(audio_data: bytes, format: str) -> float:
        """Estimate audio duration from data size."""
        # Rough estimate: ~16KB/s for WAV at 16kHz 16-bit mono
        if format == "wav":
            return len(audio_data) / 16000 * 1000
        else:  # mp3
            return len(audio_data) / 2000 * 1000  # ~2KB/s for mp3

    def save_audio(self, result: TTSResult, path: str):
        """Save TTS result to file."""
        with open(path, "wb") as f:
            f.write(result.audio_data)


# Convenience function
def speak(
    text: str,
    language: str = "en",
    emotion: str = "neutral",
    api_key: Optional[str] = None,
) -> bytes:
    """
    Quick TTS function. Returns audio bytes.
    """
    tts = MiMoTTS(api_key=api_key)
    result = tts.synthesize(text, language=language, emotion=emotion)
    return result.audio_data


if __name__ == "__main__":
    import sys

    if len(sys.argv) < 2:
        print("Usage: python mimo_tts.py 'text to speak' [language] [output.wav]")
        sys.exit(1)

    text = sys.argv[1]
    lang = sys.argv[2] if len(sys.argv) > 2 else "en"
    output = sys.argv[3] if len(sys.argv) > 3 else "output.wav"

    tts = MiMoTTS()
    result = tts.synthesize(text, language=lang)
    tts.save_audio(result, output)
    print(f"Saved {result.duration_ms:.0f}ms audio to {output} ({result.latency_ms:.0f}ms latency)")

File: mimo-voice/android/java/nie/translator/rtranslator/voice_translation/neural_networks/translation/MiMoTTS.java
/*
 * MiMo Voice — MiMo TTS (Android)
 * Replaces gTTS with MiMo V2-TTS for natural speech synthesis.
 */

package nie.translator.rtranslator.voice_translation.neural_networks.translation;

import android.content.Context;
import android.media.AudioAttributes;
import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioTrack;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;


/**
 * MiMo V2-TTS client for Android.
 *
 * Replaces gTTS (robotic) with natural, expressive speech synthesis.
 * Supports emotion, proper intonation, and natural pacing.
 *
 * Usage:
 *   MiMoTTS tts = new MiMoTTS(context, apiKey);
 *   tts.synthesize("Hello!", "en", "neutral", listener);
 */
public class MiMoTTS {
    private static final String TAG = "MiMoTTS";
    private static final String BASE_URL = "https://api.xiaomimimo.com/v1";
    private static final String MODEL = "mimo-v2-tts";
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    // Emotion presets
    public static final String EMOTION_NEUTRAL = "neutral";
    public static final String EMOTION_HAPPY = "happy";
    public static final String EMOTION_SAD = "sad";
    public static final String EMOTION_URGENT = "urgent";
    public static final String EMOTION_CALM = "calm";

    // Voice presets per language
    private static final Map<String, String[]> VOICE_PRESETS = new HashMap<>();
    static {
        VOICE_PRESETS.put("en", new String[]{"nova", "echo", "nova"});
        VOICE_PRESETS.put("ja", new String[]{"yuki", "haru", "yuki"});
        VOICE_PRESETS.put("zh", new String[]{"xiaoyi", "yunxi", "xiaoyi"});
        VOICE_PRESETS.put("ko", new String[]{"minji", "hyunwoo", "minji"});
        VOICE_PRESETS.put("es", new String[]{"elena", "carlos", "elena"});
        VOICE_PRESETS.put("fr", new String[]{"claire", "antoine", "claire"});
        VOICE_PRESETS.put("de", new String[]{"hanna", "maximilian", "hanna"});
    }

    private final Context context;
    private final String apiKey;
    private final OkHttpClient httpClient;
    private final Handler mainHandler;

    public interface TTSListener {
        void onAudioReady(byte[] audioData, String format);
        void onFailure(int[] reasons, long value);
    }

    /**
     * Initialize MiMo TTS.
     *
     * @param context Android context
     * @param apiKey  MiMo API key
     */
    public MiMoTTS(@NonNull Context context, @NonNull String apiKey) {
        this.context = context;
        this.apiKey = apiKey;
        this.mainHandler = new Handler(Looper.getMainLooper());
        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(15, TimeUnit.SECONDS)
                .readTimeout(15, TimeUnit.SECONDS)
                .build();
    }

    /**
     * Synthesize speech from text.
     *
     * @param text     Text to speak
     * @param language Language code (e.g., "en", "ja")
     * @param emotion  Emotion preset
     * @param listener Callback with audio data
     */
    public void synthesize(@NonNull String text,
                           @NonNull String language,
                           @NonNull String emotion,
                           @NonNull TTSListener listener) {
        synthesize(text, language, null, emotion, 1.0f, listener);
    }

    /**
     * Synthesize speech with full options.
     */
    public void synthesize(@NonNull String text,
                           @NonNull String language,
                           @Nullable String voice,
                           @NonNull String emotion,
                           float speed,
                           @NonNull TTSListener listener) {

        if (voice == null) {
            voice = getDefaultVoice(language);
        }

        long startTime = System.currentTimeMillis();

        JSONObject request = new JSONObject();
        try {
            request.put("model", MODEL);
            request.put("input", text);
            request.put("voice", voice);
            request.put("speed", speed);
            request.put("response_format", "wav");
            request.put("language", language);

            // Add emotion instruction
            if (!emotion.equals(EMOTION_NEUTRAL)) {
                request.put("instruction", getEmotionInstruction(emotion));
            }
        } catch (JSONException e) {
            listener.onFailure(new int[]{nie.translator.rtranslator.tools.ErrorCodes.ERROR_EXECUTING_MODEL}, 0);
            return;
        }

        Request apiRequest = new Request.Builder()
                .url(BASE_URL + "/audio/speech")
                .addHeader("Authorization", "Bearer " + apiKey)
                .addHeader("Content-Type", "application/json")
                .post(RequestBody.create(request.toString(), JSON))
                .build();

        httpClient.newCall(apiRequest).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                Log.e(TAG, "TTS call failed: " + e.getMessage());
                mainHandler.post(() ->
                        listener.onFailure(new int[]{nie.translator.rtranslator.tools.ErrorCodes.ERROR_EXECUTING_MODEL}, 0));
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                if (!response.isSuccessful()) {
                    Log.e(TAG, "TTS API error: " + response.code());
                    mainHandler.post(() ->
                            listener.onFailure(new int[]{nie.translator.rtranslator.tools.ErrorCodes.ERROR_EXECUTING_MODEL}, 0));
                    return;
                }

                byte[] audioData = response.body().bytes();
                long latency = System.currentTimeMillis() - startTime;
                Log.i(TAG, "TTS completed: " + audioData.length + " bytes in " + latency + "ms");

                mainHandler.post(() -> listener.onAudioReady(audioData, "wav"));
            }
        });
    }

    /**
     * Synthesize and play directly.
     */
    public void synthesizeAndPlay(@NonNull String text,
                                   @NonNull String language,
                                   @NonNull String emotion) {
        synthesize(text, language, emotion, new TTSListener() {
            @Override
            public void onAudioReady(byte[] audioData, String format) {
                playAudio(audioData);
            }

            @Override
            public void onFailure(int[] reasons, long value) {
                Log.e(TAG, "TTS failed");
            }
        });
    }

    /**
     * Play WAV audio data.
     */
    private void playAudio(byte[] wavData) {
        // Skip WAV header (44 bytes) for PCM playback
        int headerSize = 44;
        byte[] pcmData = new byte[wavData.length - headerSize];
        System.arraycopy(wavData, headerSize, pcmData, 0, pcmData.length);

        int sampleRate = 24000; // MiMo TTS default
        int channelConfig = AudioFormat.CHANNEL_OUT_MONO;
        int audioFormat = AudioFormat.ENCODING_PCM_16BIT;

        AudioTrack audioTrack = new AudioTrack.Builder()
                .setAudioAttributes(new AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_ASSISTANT)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                        .build())
                .setAudioFormat(new AudioFormat.Builder()
                        .setEncoding(audioFormat)
                        .setSampleRate(sampleRate)
                        .setChannelMask(channelConfig)
                        .build())
                .setBufferSizeInBytes(pcmData.length)
                .setTransferMode(AudioTrack.MODE_STATIC)
                .build();

        audioTrack.write(pcmData, 0, pcmData.length);
        audioTrack.play();

        // Release after playback
        audioTrack.setNotificationMarkerPosition(pcmData.length / 2);
        audioTrack.setPlaybackPositionUpdateListener(new AudioTrack.OnPlaybackPositionUpdateListener() {
            @Override
            public void onMarkerReached(AudioTrack track) {
                track.release();
            }
            @Override
            public void onPeriodicNotification(AudioTrack track) {}
        });
    }

    /**
     * Get default voice for language.
     */
    private String getDefaultVoice(String language) {
        String[] voices = VOICE_PRESETS.get(language);
        if (voices != null) {
            return voices[0]; // Default voice
        }
        return "nova"; // Fallback
    }

    /**
     * Get emotion instruction for TTS.
     */
    private String getEmotionInstruction(String emotion) {
        switch (emotion) {
            case EMOTION_HAPPY:
                return "Speak with a warm, happy tone.";
            case EMOTION_SAD:
                return "Speak with a gentle, somber tone.";
            case EMOTION_URGENT:
                return "Speak urgently and clearly. Important!";
            case EMOTION_CALM:
                return "Speak calmly and soothingly.";
            default:
                return "";
        }
    }

    /**
     * Release resources.
     */
    public void shutdown() {
        httpClient.dispatcher().executorService().shutdown();
    }
}
